presto-hive/src/main/java/com/facebook/presto/hive/cache/HiveCachingHdfsConfiguration.java

```diff
@@ -62,10 +62,11 @@ public HiveCachingHdfsConfiguration(
     @Override
     public Configuration getConfiguration(HdfsContext context, URI uri)
     {
+        Configuration defaultConfig = hiveHdfsConfiguration.getConfiguration(context, uri);
```

Comment:
In `HiveHdfsConfiguration.getConfiguration()` which will be eventually called from here, we update the configuration object based on the various `DynamicConfigurationProviders`. For example `GcsConfigurationProvider` update the config based on the URI. So yeah different URIs might yield different configurations and in future we might other `DynamicConfigurationProviders` as well. 
https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/HiveHdfsConfiguration.java#L64-L67

Reply:
I understand that different configuration providers would have different config, but this is called every split and I'd guess that all splits from the same connector pretty much would have the same config?

Reply:
yeah agree. But the problem is that these Configuration objects are mutable and there are some places of code where we modify & add parameters to it. So how do we make sure things are right.. if a modifiable config is being passed around and being shared ?

Reply:
there are only two DynamicConfigurationProvider subclasses and they don't set different values for different uris. So we could actually share the configurations for different uris in this case.

How about adding a new function to HdfsConfiguration and DynamicConfigurationProvider:
canShareConfiguration(uri1, uri2) 
to indicate if two uris can share the configuration?

And the (uri.compareTo(factoryUri) != 0) below can be replaced with canShareConfiguration(uri, factoryUri) .
In that case, we could share configuration for any URIs safely for the existing DynamicConfigurationProvider subclasses.

Reply:
What about extend `Configuration` and implement a copy-on-write version of it? Only copy when `set` is called.

Reply:
Haven't checked the details on whether it is feasible to do so. This may require a much larger refactoring effort. If we want to make it copy-on-write, we need to make sure
 1. If we set 10 fields together, we want to copy it once instead of 10 times.
 2. The original pointer is not stored somewhere so that there are no multiple versions of the configs used in different places.

Reply:
For identifying the first update, we can just return an immutable map by default and check whether the map is modifiable. If not, copy and update, if it's modifiable already, just update.